File: src/main/java/org/eclairjs/nashorn/wrap/sql/streaming/DataStreamWriter.java
package org.eclairjs.nashorn.wrap.sql.streaming;
/*                                                                         
* Copyright 2016 IBM Corp.                                                 
*                                                                          
* Licensed under the Apache License, Version 2.0 (the "License");          
* you may not use this file except in compliance with the License.         
* You may obtain a copy of the License at                                  
*                                                                          
*      http://www.apache.org/licenses/LICENSE-2.0                          
*                                                                          
* Unless required by applicable law or agreed to in writing, software      
* distributed under the License is distributed on an "AS IS" BASIS,        
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
* See the License for the specific language governing permissions and      
* limitations under the License.                                           
*/ 

import org.apache.spark.sql.streaming.OutputMode;
import org.eclairjs.nashorn.Utils;
import org.eclairjs.nashorn.wrap.WrappedFunction;
import org.apache.log4j.Logger;
import org.eclairjs.nashorn.wrap.WrappedClass;


public class DataStreamWriter extends WrappedClass {

 static Logger logger = Logger.getLogger(DataStreamWriter.class);

    static WrappedFunction F_outputMode = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            logger.debug("outputMode");
            Object returnValue = null;
            org.apache.spark.sql.streaming.DataStreamWriter _dataStreamWriter = (org.apache.spark.sql.streaming.DataStreamWriter) ((DataStreamWriter) thiz).getJavaObject();
            String outputMode = (String) args[0];
            returnValue = _dataStreamWriter.outputMode(outputMode);
            // return Utils.javaToJs(returnValue);
            return new org.eclairjs.nashorn.wrap.sql.streaming.DataStreamWriter((org.apache.spark.sql.streaming.DataStreamWriter)returnValue);
        }
    };

    static WrappedFunction F_trigger = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            logger.debug("trigger");
            Object returnValue = null;
            org.apache.spark.sql.streaming.DataStreamWriter _dataStreamWriter = (org.apache.spark.sql.streaming.DataStreamWriter) ((DataStreamWriter) thiz).getJavaObject();
            org.apache.spark.sql.streaming.Trigger trigger = (org.apache.spark.sql.streaming.Trigger) Utils.toObject(args[0]);
            returnValue = _dataStreamWriter.trigger(trigger);
            // return Utils.javaToJs(returnValue);
            return new org.eclairjs.nashorn.wrap.sql.streaming.DataStreamWriter((org.apache.spark.sql.streaming.DataStreamWriter)returnValue);
        }
    };

    static WrappedFunction F_queryName = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            logger.debug("queryName");
            Object returnValue = null;
            org.apache.spark.sql.streaming.DataStreamWriter _dataStreamWriter = (org.apache.spark.sql.streaming.DataStreamWriter) ((DataStreamWriter) thiz).getJavaObject();
            String queryName = (String) args[0];
            returnValue = _dataStreamWriter.queryName(queryName);
            // return Utils.javaToJs(returnValue);
            return new org.eclairjs.nashorn.wrap.sql.streaming.DataStreamWriter((org.apache.spark.sql.streaming.DataStreamWriter)returnValue);
        }
    };

    static WrappedFunction F_format = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            logger.debug("format");
            Object returnValue = null;
            org.apache.spark.sql.streaming.DataStreamWriter _dataStreamWriter = (org.apache.spark.sql.streaming.DataStreamWriter) ((DataStreamWriter) thiz).getJavaObject();
            String source = (String) args[0];
            returnValue = _dataStreamWriter.format(source);
            // return Utils.javaToJs(returnValue);
            return new org.eclairjs.nashorn.wrap.sql.streaming.DataStreamWriter((org.apache.spark.sql.streaming.DataStreamWriter)returnValue);
        }
    };

    static WrappedFunction F_partitionBy = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            logger.debug("partitionBy");
            Object returnValue = null;
            org.apache.spark.sql.streaming.DataStreamWriter _dataStreamWriter = (org.apache.spark.sql.streaming.DataStreamWriter) ((DataStreamWriter) thiz).getJavaObject();
            String colNames = (String) args[0]; // TODO: handle repeated parm 'colNames'
            returnValue = _dataStreamWriter.partitionBy(colNames);
            // return Utils.javaToJs(returnValue);
            return new org.eclairjs.nashorn.wrap.sql.streaming.DataStreamWriter((org.apache.spark.sql.streaming.DataStreamWriter)returnValue);
        }
    };

    static WrappedFunction F_option = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            logger.debug("option");
            Object returnValue = null;
            org.apache.spark.sql.streaming.DataStreamWriter _dataStreamWriter = (org.apache.spark.sql.streaming.DataStreamWriter) ((DataStreamWriter) thiz).getJavaObject();
            String key = (String) args[0];
            double value =  Utils.toDouble(args[1]);
            returnValue = _dataStreamWriter.option(key,value);
            // return Utils.javaToJs(returnValue);
            return new org.eclairjs.nashorn.wrap.sql.streaming.DataStreamWriter((org.apache.spark.sql.streaming.DataStreamWriter)returnValue);
        }
    };

    static WrappedFunction F_options = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            logger.debug("options");
            Object returnValue = null;
            org.apache.spark.sql.streaming.DataStreamWriter _dataStreamWriter = (org.apache.spark.sql.streaming.DataStreamWriter) ((DataStreamWriter) thiz).getJavaObject();
            java.util.Map options = (java.util.Map) Utils.toObject(args[0]);
            returnValue = _dataStreamWriter.options(options);
            // return Utils.javaToJs(returnValue);
            return new org.eclairjs.nashorn.wrap.sql.streaming.DataStreamWriter((org.apache.spark.sql.streaming.DataStreamWriter)returnValue);
        }
    };

    static WrappedFunction F_start = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            logger.debug("start");
            Object returnValue = null;
            org.apache.spark.sql.streaming.DataStreamWriter _dataStreamWriter = (org.apache.spark.sql.streaming.DataStreamWriter) ((DataStreamWriter) thiz).getJavaObject();
            
            if (args.length==0) {
              returnValue = _dataStreamWriter.start();
            
            } else {
            String path = (String) args[0];
              returnValue = _dataStreamWriter.start(path);
            
            }
            return Utils.javaToJs(returnValue);
            //return new org.eclairjs.nashorn.wrap.sql.streaming.StreamingQuery((org.apache.spark.sql.streaming.StreamingQuery)returnValue);
        }
    };

    static WrappedFunction F_foreach = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            logger.debug("foreach");
            Object returnValue = null;
            org.apache.spark.sql.streaming.DataStreamWriter _dataStreamWriter = (org.apache.spark.sql.streaming.DataStreamWriter) ((DataStreamWriter) thiz).getJavaObject();
            org.apache.spark.sql.ForeachWriter writer = (org.apache.spark.sql.ForeachWriter) Utils.toObject(args[0]);
            returnValue = _dataStreamWriter.foreach(writer);
            return Utils.javaToJs(returnValue);
           // return new org.eclairjs.nashorn.wrap.sql.streaming.DataStreamWriter((org.apache.spark.sql.streaming.DataStreamWriter)returnValue);
        }
    };


    private org.apache.spark.sql.streaming.DataStreamWriter _dataStreamWriter;

    public DataStreamWriter(org.apache.spark.sql.streaming.DataStreamWriter _dataStreamWriter)
    {
       logger.debug("constructor");
       this._dataStreamWriter = _dataStreamWriter;
    }

    static public String getModuleName() {
        return "sql.streaming.DataStreamWriter";
    }

    public boolean checkInstance(Object other) {
        return other instanceof DataStreamWriter;
    }

    public Object getJavaObject() {
        return _dataStreamWriter;
    }

    @Override
    public String toString() {

        return _dataStreamWriter.toString();
    }

    public String getClassName() {
        return "DataStreamWriter";
    }

    // get the value of that named property
    @Override
    public Object getMember(String name) {
        switch (name) {
            case "outputMode":
                return F_outputMode;
            case "trigger":
                return F_trigger;
            case "queryName":
                return F_queryName;
            case "format":
                return F_format;
            case "partitionBy":
                return F_partitionBy;
            case "option":
                return F_option;
            case "options":
                return F_options;
            case "start":
                return F_start;
            case "foreach":
                return F_foreach;
        }
        return super.getMember(name);
    }

    @Override
    public boolean hasMember(String name) {
        switch (name) {
            case "outputMode":
            case "trigger":
            case "queryName":
            case "format":
            case "partitionBy":
            case "option":
            case "options":
            case "start":
            case "foreach":
                return true;
        }
        return super.hasMember(name);
    }

}

File: src/main/resources/eclairjs/sql/streaming/DataStreamWriter.js
/*                                                                         
* Copyright 2016 IBM Corp.                                                 
*                                                                          
* Licensed under the Apache License, Version 2.0 (the "License");          
* you may not use this file except in compliance with the License.         
* You may obtain a copy of the License at                                  
*                                                                          
*      http://www.apache.org/licenses/LICENSE-2.0                          
*                                                                          
* Unless required by applicable law or agreed to in writing, software      
* distributed under the License is distributed on an "AS IS" BASIS,        
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
* See the License for the specific language governing permissions and      
* limitations under the License.                                           
*/ 

(function () {

    
    /**
     * @classdesc
     * 
     * Interface used to write a streaming {@link Dataset} to external storage systems (e.g. file systems,
     * key-value stores, etc). Use {@link writeStream} to access this.
     *
     * @since EclairJS 0.7 Spark  2.0.0
     * @class DataStreamWriter
     * @memberof module:eclairjs/sql/streaming
     */
    var DataStreamWriter = Java.type('org.eclairjs.nashorn.wrap.sql.streaming.DataStreamWriter');
    
    
    /**
     * 
     * Specifies how data of a streaming DataFrame/Dataset is written to a streaming sink.
     *   - `append`: only the new rows in the streaming DataFrame/Dataset will be
     *                            written to the sink
     *   - `complete`: all the rows in the streaming DataFrame/Dataset will be written
     *                              to the sink every time these is some updates
     * @function
     * @name module:eclairjs/sql/streaming.DataStreamWriter#outputMode
     * @since EclairJS 0.7 Spark  2.0.0
     * @param {string} outputMode
     * @returns {module:eclairjs/sql/streaming.DataStreamWriter} 
     */

    
    
    /**
     * 
     * Set the trigger for the stream query. The default value is `ProcessingTime(0)` and it will run
     * the query as fast as possible.
     *
     * @function
     * @name module:eclairjs/sql/streaming.DataStreamWriter#trigger
     *
     * @example 
     *   df.writeStream().trigger(ProcessingTime.create("10 seconds"))
     *
     * @since EclairJS 0.7 Spark  2.0.0
     * @param {module:eclairjs/sql/streaming.Trigger} trigger
     * @returns {module:eclairjs/sql/streaming.DataStreamWriter} 
     */
    
    
    /**
     * 
     * Specifies the name of the {@link StreamingQuery} that can be started with `start()`.
     * This name must be unique among all the currently active queries in the associated SQLContext.
     *
     * @function
     * @name module:eclairjs/sql/streaming.DataStreamWriter#queryName
     * @since EclairJS 0.7 Spark  2.0.0
     * @param {string} queryName
     * @returns {module:eclairjs/sql/streaming.DataStreamWriter} 
     */

    
    /**
     * 
     * Specifies the underlying output data source. Built-in options include "parquet", "json", etc.
     *
     * @function
     * @name module:eclairjs/sql/streaming.DataStreamWriter#format
     * @since EclairJS 0.7 Spark  2.0.0
     * @param {string} source
     * @returns {module:eclairjs/sql/streaming.DataStreamWriter} 
     */
    
    
    /**
     * Partitions the output by the given columns on the file system. If specified, the output is
     * laid out on the file system similar to Hive's partitioning scheme. As an example, when we
     * partition a dataset by year and then month, the directory layout would look like:
     *
     *   - year=2016/month=01/
     *   - year=2016/month=02/
     *
     * Partitioning is one of the most widely used techniques to optimize physical data layout.
     * It provides a coarse-grained index for skipping unnecessary data reads when queries have
     * predicates on the partitioned columns. In order for partitioning to work well, the number
     * of distinct values in each column should typically be less than tens of thousands.
     *
     * This was initially applicable for Parquet but in 1.5+ covers JSON, text, ORC and avro as well.
     *
     * @function
     * @name module:eclairjs/sql/streaming.DataStreamWriter#partitionBy
     * @since EclairJS 0.5 Spark  1.4.0
     * @param {...string} colNames
     * @returns {module:eclairjs/sql/streaming.DataStreamWriter} 
     */
    
    
    /**
     * 
     * Adds an output option for the underlying data source.
     *
     * @function
     * @name module:eclairjs/sql/streaming.DataStreamWriter#option
     * @since EclairJS 0.7 Spark  2.0.0
     * @param {string} key
     * @param {string} value
     * @returns {module:eclairjs/sql/streaming.DataStreamWriter} 
     */
    
    
    /**
     * 
     * Starts the execution of the streaming query, which will continually output results to the given
     * path as new data arrives. The returned {@link StreamingQuery} object can be used to interact with
     * the stream.
     *
     * @function
     * @name module:eclairjs/sql/streaming.DataStreamWriter#start
     * @since EclairJS 0.7 Spark  2.0.0
     * @param {string} [path]
     * @returns {module:eclairjs/sql/streaming.StreamingQuery} 
     */

    
    /**
     * Starts the execution of the streaming query, which will continually send results to the given external system.
     * processFunction as as new data arrives. The processFunction can be used to send the data
     * generated by the [[DataFrame]]/{@link module:eclairjs/sql.Dataset} to an external system.
     *
     * @function
     * @name module:eclairjs/sql/streaming.DataStreamWriter#foreach
     *
     * @example
     * var query = counts.writeStream().foreach(function(partitionId, version, webSocket) {
     *       // open connection
     *      var connection = webSocket.open(....);
     *      return connection;
     *  },[webSocket],
     *  function(connection, value) {
     *     connection.send(value);
     *   },
     *  function(connection) {
     *     connection.close();
     *  }).start();
     *  
     *
     * @since EclairJS 0.5 Spark  2.0.0
     * @param {module:eclairjs/sql/streaming.DataStreamWriter~openCallback} openCallback Used to open connection to external system.
     * @param {module:eclairjs/sql/streaming.DataStreamWriter~processCallback} processCallback use to send the data external system.
     * @param {module:eclairjs/sql/streaming.DataStreamWriter~closeCallback} closeFunction Used to close connection to external system.
     * @param {object[]} [openFunctionBindArgs]
     * @param {object[]} [processFunctionBindArgs]
     * @param {object[]} [closeFunctionBindArgs]
     * @returns {module:eclairjs/sql/streaming.DataStreamWriter}
     */

    /**
     * This callback Used to open connection to external system.
     * @callback module:eclairjs/sql/streaming.DataStreamWriter~openCallback
     * @param {number} partitionId
     * @param {number} version
     * @param {object[]} [bindArgs]
     * @returns {object} connection that is passed to {@link module:eclairjs/sql/streaming.DataStreamWriter~processCallback}
     * and {@link module:eclairjs/sql/streaming.DataStreamWriter~closeCallback}
     */

    /**
     * This callback consume data generated by a StreamingQuery. Typically this is used to send the generated data
     * to external systems from each partition so you usually should do all the initialization (e.g. opening a connection
     * or initiating a transaction) in the {@link module:eclairjs/sql/streaming.DataStreamWriter~openCallback}.
     * @callback module:eclairjs/sql/streaming.DataStreamWriter~processCallback
     * @param {object} connection
     * @param {number} value
     * @param {object[]} [bindArgs]
     */

    /**
     * Used to open connection to external system.
     * @callback module:eclairjs/sql/streaming.DataStreamWriter~closeCallback
     * @param {object} connection
     * @param {object[]} [bindArgs]
     */
    
    module.exports = DataStreamWriter;
})();